tests/tape/tapes/test_tape.py

```diff
@@ -979,3 +981,161 @@ def test_multiple_output_values(self, tol):
+            qml.BasisState(np.array([1, 0]), wires=[0, 1])
+            qml.RY(0.5, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0) @ qml.PauliY(1))
+
+        copied_tape = tape.copy()
+
+        assert copied_tape is not tape
+
+        # the operations are simply references
+        assert copied_tape.operations == tape.operations
+        assert copied_tape.observables == tape.observables
+        assert copied_tape.measurements == tape.measurements
+        assert copied_tape.operations[0] is tape.operations[0]
+
+        # operation data is also a reference
+        assert copied_tape.operations[0].wires is tape.operations[0].wires
+        assert copied_tape.operations[0].data[0] is tape.operations[0].data[0]
+
+        # check that all tape metadata is identical
+        assert tape.get_parameters() == copied_tape.get_parameters()
+        assert tape.wires == copied_tape.wires
+        assert tape.data == copied_tape.data
+
+        # since the copy is shallow, mutating the parameters
+        # on one tape will affect the parameters on another tape
+        new_params = [np.array([0, 0]), 0.2]
+        tape.set_parameters(new_params)
+
+        # check that they are the same objects in memory
+        for i, j in zip(tape.get_parameters(), new_params):
+            assert i is j
+
+        for i, j in zip(copied_tape.get_parameters(), new_params):
+            assert i is j
+
+    @pytest.mark.parametrize("copy_fn", [lambda tape: tape.copy(copy_operations=True), lambda tape: copy.copy(tape)])
+    def test_shallow_copy_with_operations(self, copy_fn):
+        """Test that shallow copying of a tape and operations allows
+        parameters to be set independently"""
+
+        with QuantumTape() as tape:
+            qml.BasisState(np.array([1, 0]), wires=[0, 1])
+            qml.RY(0.5, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0) @ qml.PauliY(1))
+
+        copied_tape = copy_fn(tape)
+
+        assert copied_tape is not tape
+
+        # the operations are not references; they are unique objects
+        assert copied_tape.operations != tape.operations
+        assert copied_tape.observables != tape.observables
+        assert copied_tape.measurements != tape.measurements
+        assert copied_tape.operations[0] is not tape.operations[0]
+
+        # however, the underlying operation data *is still shared*
+        assert copied_tape.operations[0].wires is tape.operations[0].wires
+        assert copied_tape.operations[0].data[0] is tape.operations[0].data[0]
+
+        assert tape.get_parameters() == copied_tape.get_parameters()
+        assert tape.wires == copied_tape.wires
+        assert tape.data == copied_tape.data
+
+        # Since they have unique operations, mutating the parameters
+        # on one tape will *not* affect the parameters on another tape
+        new_params = [np.array([0, 0]), 0.2]
+        tape.set_parameters(new_params)
```

Comment:
I wonder if we'll ever hit a point where this nonstandard copying behaviour impacts developers or power users?

Reply:
Good question. By default, `tape.copy()` will perform a single shallow copy --- as users or developers would expect. It is only when `copy_operations=True` (so something the user/developer will have to explicitly request) does the second shallow copy of operations take place. Since this is not default behaviour, I'm not too worried.

I'm more worried about the non-standard behaviour of _deep copying_. Users/developers expect deep copy to copy _everything_; however the parameters are not copied (since PyTorch doesn't allow this). I think fixing this is a good thing; it's easy to get everything working with `copy.deepcopy(tape)`, and then realize only later that while autograd and TF work, Torch doesn't.

Perhaps the deep copy could _try_ and copy the parameters, and only skip them if an exception occurs (e.g., Torch)?